sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala

```diff
@@ -55,61 +57,133 @@ private[hive] class HiveFunctionRegistry(
         } else if (
-          classOf[AbstractGenericUDAFResolver].isAssignableFrom(functionInfo.getFunctionClass)) {
-          val udaf = HiveUDAFFunction(name, new HiveFunctionWrapper(functionClassName), children)
-          udaf.dataType // Force it to check input data types.
+          classOf[AbstractGenericUDAFResolver].isAssignableFrom(hiveUDFClass)) {
+          val udaf = HiveUDAFFunction(name, hiveUDFWrapper, children)
+          if (udaf.resolved) {
+            udaf.dataType // Force it to check input data types.
+          }
           udaf
-        } else if (classOf[UDAF].isAssignableFrom(functionInfo.getFunctionClass)) {
+        } else if (classOf[UDAF].isAssignableFrom(hiveUDFClass)) {
           val udaf = HiveUDAFFunction(
-            name, new HiveFunctionWrapper(functionClassName), children, isUDAFBridgeRequired = true)
-          udaf.dataType  // Force it to check input data types.
+            name, hiveUDFWrapper, children, isUDAFBridgeRequired = true)
+          if (udaf.resolved) {
+            udaf.dataType  // Force it to check input data types.
+          }
           udaf
-        } else if (classOf[GenericUDTF].isAssignableFrom(functionInfo.getFunctionClass)) {
-          val udtf = HiveGenericUDTF(name, new HiveFunctionWrapper(functionClassName), children)
-          udtf.elementTypes // Force it to check input data types.
+        } else if (classOf[GenericUDTF].isAssignableFrom(hiveUDFClass)) {
+          val udtf = HiveGenericUDTF(name, hiveUDFWrapper, children)
+          if (udtf.resolved) {
+            udtf.elementTypes // Force it to check input data types.
+          }
           udtf
         } else {
-          throw new AnalysisException(s"No handler for udf ${functionInfo.getFunctionClass}")
+          throw new AnalysisException(s"No handler for udf ${hiveUDFClass}")
         }
       } catch {
         case analysisException: AnalysisException =>
-          // If the exception is an AnalysisException, just throw it.
           throw analysisException
         case throwable: Throwable =>
-          // If there is any other error, we throw an AnalysisException.
-          val errorMessage = s"No handler for Hive udf ${functionInfo.getFunctionClass} " +
+          val errorMessage = s"No handler for Hive udf ${hiveUDFClass} " +
             s"because: ${throwable.getMessage}."
           throw new AnalysisException(errorMessage)
       }
     }
+    builder
+  }
+
+  override def lookupFunction(name: String, children: Seq[Expression]): Expression = {
+    val builder = underlying.lookupFunctionBuilder(name)
+    if (builder.isDefined) {
+      builder.get(children)
+    } else {
+      // We only look it up to see if it exists, but do not include it in the HiveUDF since it is
+      // not always serializable.
+      val optFunctionInfo = Option(getFunctionInfo(name.toLowerCase))
+      if (optFunctionInfo.isEmpty) {
+        val catalogFunc = loadHivePermanentFunction(name).getOrElse(
```

Comment:
Because we main hive permanent functions in Spark too, when `getFunctionInfo` returns null, we can get `CatalogFunction` through catalog and load necessary jars, files. Then we create `FunctionBuilder` as the same as temporary function.